# PN30 - dynamic relational flip ridge

## tl;dr

On a fresh 500-number odd interval, restoring singularity-driven AB/BA flips raised the difficult unresolved-composite AUC from `0.5301` to `0.5663`. The frozen one-sided result was `p=0.06199`, so the direction is suggestive but not statistically reliable. Post-hoc decomposition shows that the possible gain came from stronger cancellation among signed child orientations, not from different individual pair magnitudes.


## Context & Methods

For each child wave, normalized phase is `theta=(N mod w)/w`. In each unordered pair, the smaller theta becomes Phase A. The dynamically oriented pair coordinate is reflected around 1.0 whenever orientation flips. Three signed pair coordinates are averaged, and their displacement is halved at each of two upward rungs.

### Key assumptions

- Child pairs are `{1,13}`, `{3,11}`, `{5,9}`.
- PN29's completion function is retained unchanged.
- Coordinates for odd integers 1001-1999 were frozen before labels.
- Prime labels use direct trial division after freeze; no sieve is used.
- Post-hoc cancellation diagnostics are descriptive and not frozen endpoints.


In [1]:
import csv
import json
from pathlib import Path

HERE = Path.cwd()
results = json.loads((HERE / 'PN30_DYNAMIC_RELATIONAL_FLIP_RESULTS.json').read_text(encoding='utf-8'))
validation = json.loads((HERE / 'PN30_DYNAMIC_RELATIONAL_FLIP_VALIDATION.json').read_text(encoding='utf-8'))
posthoc = json.loads((HERE / 'PN30_DYNAMIC_RELATIONAL_FLIP_POSTHOC.json').read_text(encoding='utf-8'))
with (HERE / 'PN30_DYNAMIC_RELATIONAL_FLIP_SCORED.csv').open(encoding='utf-8', newline='') as handle:
    rows = list(csv.DictReader(handle))
print('status:', results['status'])
print('validation:', validation['checks_passed'], '/', validation['checks_total'])
assert validation['all_checks_passed'] is True
assert len(rows) == 500


status: CHILD-FILTER SUPPORT ONLY
validation: 7 / 7


## Data

In [2]:
print(results['population'])
print('orientation counts:', results['orientation_counts'])
assert results['population']['sieve_used'] is False
assert results['population']['prime_n'] == 135
assert results['population']['odd_composite_n'] == 365
assert results['population']['unresolved_composite_n'] == 90


{'n': 500, 'prime_n': 135, 'odd_composite_n': 365, 'unresolved_composite_n': 90, 'range': 'odd integers 1001 through 1999 inclusive', 'sieve_used': False, 'label_method': 'direct trial division after coordinate freeze'}
orientation counts: {'1_13': {'tie': 39, '1->13': 461}, '3_11': {'11->3': 183, '3->11': 302, 'tie': 15}, '5_9': {'5->9': 267, '9->5': 222, 'tie': 11}}


## Results - dynamic coordinate

In [3]:
overall = results['dynamic']['overall_prime_vs_odd_composite']
unresolved = results['dynamic']['prime_vs_unresolved_composite']
print('overall:', overall)
print('unresolved:', unresolved)
assert overall['auc_prime_more_ridge_close'] > 0.78
assert overall['permutation']['one_sided_p'] < 0.01
assert 0.56 < unresolved['auc_prime_more_ridge_close'] < 0.57
assert unresolved['permutation']['one_sided_p'] > 0.05


overall: {'prime_n': 135, 'composite_n': 365, 'prime_mean_distance': 0.07595126369461597, 'composite_mean_distance': 0.1453213321394466, 'prime_median_distance': 0.057136298045389, 'composite_median_distance': 0.140207986508176, 'auc_prime_more_ridge_close': 0.7838863521055301, 'permutation': {'statistic': 'mean_composite_distance_minus_mean_prime_distance', 'alternative': 'prime distance is lower', 'observed': 0.06937006844483062, 'permutations': 10000, 'seed': 30001, 'null_mean': 4.715782425246041e-05, 'one_sided_p': 9.999000099990002e-05}}
unresolved: {'prime_n': 135, 'composite_n': 90, 'prime_mean_distance': 0.07595126369461597, 'composite_mean_distance': 0.08677071584859332, 'prime_median_distance': 0.057136298045389, 'composite_median_distance': 0.103765151297516, 'auc_prime_more_ridge_close': 0.5662551440329218, 'permutation': {'statistic': 'mean_composite_distance_minus_mean_prime_distance', 'alternative': 'prime distance is lower', 'observed': 0.010819452153977352, 'permutatio

## Results - same-interval static comparator

In [4]:
static = results['static_same_interval_control']
delta = results['dynamic_minus_static_auc']
print('static:', static)
print('dynamic minus static AUC:', delta)
assert delta['overall'] < 0
assert delta['unresolved'] > 0.03


static: {'overall_prime_vs_odd_composite': {'prime_n': 135, 'composite_n': 365, 'prime_mean_distance': 0.008931863341882297, 'composite_mean_distance': 0.08424213405002363, 'prime_median_distance': 0.00904069492965726, 'composite_median_distance': 0.0794690303354418, 'auc_prime_more_ridge_close': 0.8677016742770167}, 'prime_vs_unresolved_composite': {'prime_n': 135, 'composite_n': 90, 'prime_mean_distance': 0.008931863341882297, 'composite_mean_distance': 0.008992938070815524, 'prime_median_distance': 0.00904069492965726, 'composite_median_distance': 0.009116165582445886, 'auc_prime_more_ridge_close': 0.5301234567901234}}
dynamic minus static AUC: {'overall': -0.08381532217148657, 'unresolved': 0.03613168724279836}


## Results - post-hoc signed cancellation mechanism

In [5]:
print('prime:', posthoc['prime'])
print('unresolved composite:', posthoc['unresolved_composite'])
assert posthoc['status'].startswith('POST-HOC')
assert posthoc['prime']['mean_signed_cancellation_fraction'] > posthoc['unresolved_composite']['mean_signed_cancellation_fraction']


prime: {'n': 135, 'mean_child_rung_collapsed_distance': 0.30380505477846387, 'median_child_rung_collapsed_distance': 0.228545192181556, 'mean_of_pair_absolute_distances': 0.6071560247961011, 'mean_signed_cancellation_fraction': 0.49963872121399, 'median_signed_cancellation_fraction': 0.625026903706254}
unresolved composite: {'n': 90, 'mean_child_rung_collapsed_distance': 0.3470828633943733, 'median_child_rung_collapsed_distance': 0.415060605190064, 'mean_of_pair_absolute_distances': 0.6074003237118341, 'mean_signed_cancellation_fraction': 0.42846947216287146, 'median_signed_cancellation_fraction': 0.3145575918755631}


## Takeaways

1. PN30 restores the singularity flip that PN29 omitted.
2. Dynamic orientation weakened the easy overall child-divisor screen but improved the hard unresolved comparison by `+0.0361` AUC.
3. The improvement missed the frozen significance threshold (`p=0.06199`) and requires replication.
4. Pair magnitudes were nearly identical; the descriptive difference appeared in how their signed orientations cancelled.
5. This is a reversible ARA coordinate diagnostic, not a prime generator or certification method.
